File: 프로그래머스/lv2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool visit[8];
int cnt;

void dfs(int l, vector<vector<int>> dungeons) {
    int tmp = count(visit, visit + dungeons.size(), true);
    if (tmp > cnt)
        cnt = tmp;
    for (int i = 0; i < dungeons.size(); i++) {
        if (!visit[i] && l >= dungeons[i][0]) {
            visit[i] = true;
            dfs(l - dungeons[i][1], dungeons);
            visit[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    dfs(k, dungeons);
    return cnt;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
import static java.util.stream.Collectors.*;

class Solution {
    
    private boolean[] visit;
    private int minValue = 0;
    
    private void dfs(int[][] dungeons, int r, int depth) {
        if (depth > minValue) minValue = depth;
        for (int i = 0; i < dungeons.length; i++) {
            if (!visit[i] && r >= dungeons[i][0]) {
                visit[i] = true;
                dfs(dungeons, r - dungeons[i][1], depth + 1);
                visit[i] = false;
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        visit = new boolean[dungeons.length];
        dfs(dungeons, k, 0);
        return minValue;
    }
}